codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java

```diff
@@ -113,7 +120,194 @@ public static boolean isPseudoHeader(CharSequence header) {
+                array[offset + 3] == 't' && array[offset + 4] == 'h') {
+                return PATH;
+            }
+            return null;
+        }
+
+        private static PseudoHeaderName getMethodSchemeStatusIfEqualsNoPrefix(byte[] array, int offset) {
+            byte first = array[offset + 1];
+            if (first == 'm') {
+                if (array[offset + 2] == 'e' && array[offset + 3] == 't' && array[offset + 4] == 'h' &&
+                    array[offset + 5] == 'o' && array[offset + 6] == 'd') {
+                    return METHOD;
+                }
+            } else if (first == 's') {
+                if (array[offset + 2] == 'c' && array[offset + 3] == 'h' && array[offset + 4] == 'e' &&
+                    array[offset + 5] == 'm' && array[offset + 6] == 'e') {
+                    return SCHEME;
+                } else if (array[offset + 2] == 't' && array[offset + 3] == 'a' && array[offset + 4] == 't' &&
+                           array[offset + 5] == 'u' && array[offset + 6] == 's') {
+                    return STATUS;
+                }
+            }
+            return null;
+        }
+
+        private static PseudoHeaderName getAuthorityIfEqualsNoPrefix(byte[] array, int offset) {
+            if (array[offset + 1] == 'a' && array[offset + 2] == 'u' && array[offset + 3] == 't' &&
+                array[offset + 4] == 'h' && array[offset + 5] == 'o' && array[offset + 6] == 'r' &&
+                array[offset + 7] == 'i' && array[offset + 8] == 't' && array[offset + 9] == 'y') {
+                return AUTHORITY;
+            }
+            return null;
+        }
+
+        private static PseudoHeaderName getProtocolIfEqualsNoPrefix(byte[] array, int offset) {
+            if (array[offset + 1] == 'p' && array[offset + 2] == 'r' && array[offset + 3] == 'o' &&
+                array[offset + 4] == 't' && array[offset + 5] == 'o' && array[offset + 6] == 'c' &&
+                array[offset + 7] == 'o' && array[offset + 8] == 'l') {
+                return PROTOCOL;
+            }
+            return null;
+        }
+
+        /**
+         * Returns the {@link PseudoHeaderName} corresponding to the specified header name.
+         *
+         * @return corresponding {@link PseudoHeaderName} if any, {@code null} otherwise.
+         */
+        public static PseudoHeaderName getPseudoHeader(AsciiString header) {
+            int length = header.length();
+            if (length == 0) {
+                return null;
+            }
+            byte[] array = header.array();
+            int offset = header.arrayOffset();
+            if (array[offset] == PSEUDO_HEADER_PREFIX_BYTE) {
+                switch (length) {
+                case 5:
+                    if (header == PATH.value()) {
+                        return PATH;
+                    }
+                    // :path
+                    return getPathIfEqualsNoPrefix(array, offset);
```

Comment:
How often will these specialized paths be used? In my minds eye we can get the majority of the benefit from a `.equals(..)` and maybe some improvements to `AsciiString.equals(..)`.
We could probably speed up `AsciiString.equals(..)` a little bit. We could start off with a reference equality check. Second, the terminal `&&` sequence could skip the hash code checks if the hash code hasn't been computed: it's intended to be a short circuit to avoid iteration but if the hash code needs to be computed anyway it's no benefit and actually a big detriment since to compute the hash code we have to traverse the entire array. The `AsciiString.hashCode(..)` computation looks like it was a dominator in your flame graph.

Reply:
I like the idea of @bryce-anderson

Reply:
> How often will these specialized paths be used?

If you mean the content equality ones, in 2 cases:
1. malformed/fake pseudo-headers eg `:patc` this will go through `getPathIfEqualsNoPrefix` and similar
2. non-constant known pseudo-header(s) eg new AsciiString(":path"), will still hit `getPathIfEqualsNoPrefix` and similar

Both seems unlikely path, or at least, with this PR the last one shouldn't be "hot" anymore.
I don't remember by heart the Http 2 spec but AFAIK they should be populated by using the static table only, surely @normanmaurer remember...

> We could probably speed up AsciiString.equals(..) a little bit.

I will look at it as well, based on these insights, thanks!
Related what I could do on the slow-paths is to simplify it using `AsciiString`'s equals calls still retaining the ref equality fast paths out of such, in case of length-collisions. If hash code would happen...it is what it is, see below...

> if the hash code needs to be computed anyway it's no benefit and actually a big detriment since to compute the hash code we have to traverse the entire array.

Right, I'm aware that the existing code using `getPseudoheader` will either do it because of:
- Netty internal header name validation on HPack decoding (which happen after already inserting it on http headers)
- Netty internal header name validation on Http Header insertion (both user code and on HPack decoding see https://github.com/netty/netty/blob/ee9bb63aa5382bf9930eb0df4b690286173f126a/codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java#L559)

In both cases the `AsciiString's` hash code won't be "thrown away", so maybe I could just ignore that it happens, and re-set my expectations for the slow-paths performance (which I shouldn't care that much, effectively, unless I'm very wrong related the HTTP spec).